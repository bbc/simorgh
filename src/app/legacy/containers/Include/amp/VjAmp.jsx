import React, { useContext } from 'react';
import styled from '@emotion/styled';
import { shape, string, node } from 'prop-types';
import {
  GEL_SPACING,
  GEL_SPACING_DBL,
} from '#psammead/gel-foundations/src/spacings';
import { getSansBold } from '#psammead/psammead-styles/src/font-styles';
import { GEL_PICA } from '#psammead/gel-foundations/src/typography';
import { Helmet } from 'react-helmet';
import { GridItemMedium } from '#components/Grid';
import { ServiceContext } from '../../../../contexts/ServiceContext';

const KEYLINE_WIDTH = '0.125rem';

const AmpHead = () => (
  <Helmet>
    <script
      async
      custom-element="amp-iframe"
      src="https://cdn.ampproject.org/v0/amp-iframe-0.1.js"
    />
  </Helmet>
);

const AmpIframe = ({ children, className, width, height, src }) => (
  <amp-iframe
    class={className}
    width={width}
    height={height}
    layout="responsive"
    sandbox="allow-scripts allow-same-origin allow-top-navigation-by-user-activation allow-forms"
    resizable=""
    src={src}
  >
    {children}
  </amp-iframe>
);

const StyledAmpIframe = styled(AmpIframe)`
  > [overflow] {
    background: linear-gradient(
      0deg,
      #fff 0%,
      rgba(255, 255, 255, 1) 75%,
      rgba(255, 255, 255, 0) 100%
    );
    display: flex;
    justify-content: center;

    &::after {
      background-color: #fff;
      border-top: ${KEYLINE_WIDTH} solid ${props => props.theme.palette.EBON};
      content: '';
      display: block;
      height: 50%;
      left: 0;
      position: absolute;
      top: 50%;
      width: 100%;
      z-index: -10;
    }

    button {
      ${({ service }) => getSansBold(service)}
      ${GEL_PICA}
      background-color: ${props => props.theme.palette.EBON};
      border: 0.0625rem solid transparent;
      color: #fff;
      cursor: pointer;
      display: block;
      padding: ${GEL_SPACING} ${GEL_SPACING_DBL};
      &:hover,
      &:focus {
        text-decoration: underline;
      }
    }
  }
`;

const VjAmp = ({ ampMetadata: { imageWidth, imageHeight, image, src } }) => {
  const { service } = useContext(ServiceContext);
  return (
    <>
      <AmpHead />
      <GridItemMedium>
        <StyledAmpIframe
          width={imageWidth}
          height={imageHeight}
          src={src}
          service={service}
        >
          <div overflow="">
            <button type="button">Show more</button>
          </div>
          <amp-img layout="fill" src={image} placeholder />
        </StyledAmpIframe>
      </GridItemMedium>
    </>
  );
};

AmpIframe.propTypes = {
  children: node.isRequired,
  width: string.isRequired,
  height: string.isRequired,
  src: string.isRequired,
  className: string.isRequired,
};

VjAmp.propTypes = {
  ampMetadata: shape({
    imageWidth: string.isRequired,
    imageHeight: string.isRequired,
    image: string.isRequired,
    src: string.isRequired,
  }).isRequired,
};

export default VjAmp;
