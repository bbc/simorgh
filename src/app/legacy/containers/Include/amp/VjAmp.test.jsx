import React from 'react';
import {
  render,
  waitFor,
} from '../../../../components/react-testing-library-with-providers';
import VjAmp from './VjAmp';

const vjProps = {
  src: 'https://news.files.bbci.co.uk/include/newsspec/21841-green-diet/gahuza/app/amp?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png',
  image:
    'https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png',
  imageHeight: '360',
  imageWidth: '640',
};
describe('VJ include container on Amp', () => {
  it('should render an amp-iframe with an image placeholder', async () => {
    const { container } = render(<VjAmp ampMetadata={vjProps} />);
    expect(container.querySelector('amp-iframe')).toBeInTheDocument();
    expect(container.querySelector('amp-img')).toBeInTheDocument();
    expect(container).toMatchSnapshot();
  });

  it(`should add amp-iframe extension script to page head`, async () => {
    render(<VjAmp ampMetadata={vjProps} />);

    await waitFor(() => {
      const scripts = Array.from(document.querySelectorAll('head script'));

      expect(scripts).toEqual(
        expect.arrayContaining([
          expect.objectContaining({
            src: `https://cdn.ampproject.org/v0/amp-iframe-0.1.js`,
          }),
        ]),
      );

      expect(scripts).toHaveLength(1);
    });
  });
});
