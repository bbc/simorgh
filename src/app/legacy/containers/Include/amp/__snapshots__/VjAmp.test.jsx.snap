// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`VJ include container on Amp should render an amp-iframe with an image placeholder 1`] = `
@media (max-width: 14.9375rem) {
  .emotion-0 {
    padding: 0 0.5rem;
    margin-left: 0%;
  }
}

@media (min-width: 15rem) and (max-width: 24.9375rem) {
  .emotion-0 {
    padding: 0 0.5rem;
    margin-left: 0%;
  }
}

@media (min-width: 25rem) and (max-width: 37.4375rem) {
  .emotion-0 {
    padding: 0 1rem;
    margin-left: 0%;
  }
}

@media (min-width: 37.5rem) and (max-width: 62.9375rem) {
  .emotion-0 {
    padding: 0 1rem;
    margin-left: 0%;
  }
}

@media (min-width: 63rem) and (max-width: 79.9375rem) {
  .emotion-0 {
    margin-left: 20%;
  }
}

@media (min-width: 80rem) {
  .emotion-0 {
    margin-left: 40%;
  }
}

@supports (display: grid) {
  .emotion-0 {
    display: block;
    width: initial;
    margin: 0;
  }

  @media (max-width: 14.9375rem) {
    .emotion-0 {
      grid-template-columns: repeat(6, 1fr);
      grid-column-end: span 6;
      padding: 0 0.5rem;
      grid-column-start: 1;
    }
  }

  @media (min-width: 15rem) and (max-width: 24.9375rem) {
    .emotion-0 {
      grid-template-columns: repeat(6, 1fr);
      grid-column-end: span 6;
      padding: 0 0.5rem;
      grid-column-start: 1;
    }
  }

  @media (min-width: 25rem) and (max-width: 37.4375rem) {
    .emotion-0 {
      grid-template-columns: repeat(6, 1fr);
      grid-column-end: span 6;
      padding: 0 1rem;
      grid-column-start: 1;
    }
  }

  @media (min-width: 37.5rem) and (max-width: 62.9375rem) {
    .emotion-0 {
      grid-template-columns: repeat(5, 1fr);
      grid-column-end: span 5;
      padding: 0 1rem;
      grid-column-start: 1;
    }
  }

  @media (min-width: 63rem) and (max-width: 79.9375rem) {
    .emotion-0 {
      grid-template-columns: repeat(5, 1fr);
      grid-column-end: span 5;
      grid-column-start: 2;
    }
  }

  @media (min-width: 80rem) {
    .emotion-0 {
      grid-template-columns: repeat(10, 1fr);
      grid-column-end: span 10;
      grid-column-start: 5;
    }
  }
}

.emotion-2>[overflow] {
  background: linear-gradient(
        0deg,
        #fff 0%,
        rgba(255, 255, 255, 1) 75%,
        rgba(255, 255, 255, 0) 100%
      );
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: center;
  -ms-flex-pack: center;
  -webkit-justify-content: center;
  justify-content: center;
}

.emotion-2>[overflow]::after {
  background-color: #fff;
  border-top: 0.125rem solid #222222;
  content: '';
  display: block;
  height: 50%;
  left: 0;
  position: absolute;
  top: 50%;
  width: 100%;
  z-index: -10;
}

.emotion-2>[overflow] button {
  font-size: 0.9375rem;
  line-height: 1.25rem;
  background-color: #222222;
  border: 0.0625rem solid transparent;
  color: #fff;
  cursor: pointer;
  display: block;
  padding: 0.5rem 1rem;
}

@media (min-width: 20rem) {
  .emotion-2>[overflow] button {
    font-size: 1rem;
    line-height: 1.25rem;
  }
}

.emotion-2>[overflow] button:hover,
.emotion-2>[overflow] button:focus {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

<div>
  <div
    class="emotion-0 emotion-1"
    dir="ltr"
  >
    <amp-iframe
      class="emotion-2 emotion-3"
      height="360"
      layout="responsive"
      resizable=""
      sandbox="allow-scripts allow-same-origin allow-top-navigation-by-user-activation allow-forms"
      src="https://news.files.bbci.co.uk/include/newsspec/21841-green-diet/gahuza/app/amp?responsive=true&newsapps=true&app-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png&app-clickable=true&amp-clickable=true&amp-image-height=360&amp-image-width=640&amp-image=https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png"
      width="640"
    >
      <div
        overflow=""
      >
        <button
          type="button"
        >
          Show more
        </button>
      </div>
      <amp-img
        layout="fill"
        placeholder="true"
        src="https://news.files.bbci.co.uk/vj/live/idt-images/image-slider-asdf/app_launcher_ws_640_7ania.png"
      />
    </amp-iframe>
  </div>
</div>
`;
